controllers/grid/users/author/AuthorGridHandler.inc.php

```diff
@@ -375,7 +380,7 @@ public function updateAuthor($args, $request)
         $authorId = (int) $request->getUserVar('authorId');
         $publication = $this->getPublication();
 
-        $author = Services::get('author')->get($authorId);
+        $author = Repo::author()->get($authorId);
```

Comment:
@defstat, I know this isn't introduced by your changes, but this is a case where we get a database ID from the request without doing any checking (e.g. to ensure that the author record belongs to the current journal). Could you add a check?

(@NateWr, FYI, I'll be thinking of a way to add a good shorthand for this as part of `get` without complicating anything overmuch.)

Reply:
I like the idea of making use of callbacks more -- especially to make an API extensible for one-off use-cases. However, I think there are a few things to consider here:

1. The collector is often constructed in the outer scope. In the API, it's common to do this:

```php
$params = $slimRequest->getQueryParams();
$collector = Repo::user()->getCollector();
if ($params['userIds']) {
    $collector->filterByUserIds($params['userIds']);
}
// Lots more params checks here...
$users = Repo::getMany($collector);
```

With the anonymous function, we'd have to move the params checks into the anonymous methods, and pass any dependencies we need into the anonymous function:

```php
$users = Repo::getMany(function(Collector $c) use ($params) {
    if ($params['userIds']) {
        $collector->filterByUserIds($params['userIds']);
    }
    // Lots more params checks here...
});
```

This would cause problems when we want to reuse the code that builds up a collector, for example when submission API params are converted into a collector that is used across multiple API ops: https://github.com/pkp/pkp-lib/blob/main/api/v1/submissions/PKPSubmissionHandler.inc.php#L252-L327

2. One of the things that `Repo::user()->getCollector()` does is abstract away the distinction between the pkp-lib and app layers. If we used an anonymous function like `fn(Collector $c) =>`, the calling code would have to know whether it expects `APP\users\Collector` or `PKP\users\Collector`. If it type-hints the `PKP\users\Collector` just to be safe, the code editor will not be able to tell the developer about methods available on `APP\users\Collector`.

3. If our goal is to prevent lazy usage of something like this:

```php
$author = Repo::author()->get((int) $request->getUserVar('authorId'));
```

I'm not sure that this is simple enough to prevent the lazy use:

```php
$author = Repo::author()->get(fn(Collector $c) => $c->authorIds([$authorId])->submissionIds([$submission->getId()]));
```

Although it can be expressed in a one-liner, it is syntactically just as complicated as the `getMany()` call. I'd argue it's harder for a new dev to discover and just as prone to misuse like this:

```php
$author = Repo::author()->get(fn(Collector $c) => $c->authorIds([$authorId]));
```

For me, it seems much harder to spot a correct vs incorrect usage of this syntax during code review. Would you notice that the submission IDs are missing in the above line?

If the end goal is to prevent insecure misuse and to simplify a syntax like `getMany(...)->first()`, I can think of a couple of strategies.

One way is to implement a `getFirst()` method that is exactly like the `getMany()` method. The problem with `getMany(...)->first()` is that it is bad for performance. The SQL query returns all requested rows. The `->first()` is called on the resulting collection. A `getFirst()` method could add the necessary `->limit(1)` to the collector and return the first matching result.

Another way is to implement named methods where appropriate: `getInSubmission($id)` or `getInContext($id)`, etc. While being a little more verbose in the `Repository`, I think that this approach makes it _very easy_ to catch mistakes during code review.

Finally, in a lot of cases maybe we should be rethinking how we get authorized objects. In the example you provide, the `Submission` object has already been instantiated and authorized. Maybe we could use helper methods to get the author records directly without another end-run to the database:

```php
$author = Repo::author()->getFromSubmission($authorId, $submission);
```

This kind of pattern certainly works well in an API-like environment, where attached objects are usually in descendent paths (`/submissions/1/authors/2`) and the parent objects are already authorized and instantiated. It is less useful in situations where the objects are not attached to each other, like issues, but in such cases most of the objects ought to be available through the authorized objects.

Reply:
Good points -- and I think a helper method solves all the above issues with a lot less convolution. We would just need to watch for calls to `get` that would better be implemented using the helper.